### PR TITLE
[Enhancement] window function binary search optimization

### LINKED ISSUE #7904: [Enhancement] window function binary search optimization
## Enhancement

In the processing of window function, we use binary search to find the next partition boundaries, and it is good in most cases. But the performance will dramaticly decrease in the case of high cardinality of partition column. Think about the exteme case, every partition has only one row, then the procedure of binary search will be performed as many times as number of rows. The overhead is `O(Nlog(N))`

The optimization is straightforward, binary search will downgrade to simple tight loop when average partition size is as small as possible, and the the overhead of loop is `O(N)`

### PR DESCRIPTION
**Pull request by liuyehcf**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #7904

## Enhancement

Binary search will be replaced with pre-computation when cardinality of partition is high. But how to determine whether the cardinality is high or not? 

**Overhead analysis**

* Binary Search: `O(N/S logN)`, where `S` denote the average partition size
* Pre-computation：`O(N)`

The default `chunk_size` is set to `4096`, let `N = 4096`, the overhead of single binary search is `log 4096 = 12`。

So, theoretically, regardless of compiler optimizations, binary search is better when `S > 12` , otherwise pre-computation is better. And considering the error of estimation and for simplicity, we set the threshould to `8`, blow which is considered as high cardinality.

## Test

* 1fe/3be, 64c/128g

```sql
-- Q1, ssb-100g, dop=8
SELECT max(wx) FROM ( SELECT avg(LO_TAX) OVER (PARTITION BY LO_ORDERKEY,LO_PARTKEY ORDER BY LO_TAX ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING) AS wx FROM lineorder_flat )a;

-- Q2, ssb-1g, dop=1
SELECT max(wx) FROM (
SELECT avg(LO_TAX) OVER (PARTITION BY C_REGION,C_NATION,C_CITY,S_CITY,S_NATION,S_REGION,P_TYPE ORDER BY LO_ORDERKEY ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING) AS wx FROM lineorder_flat
)a;

-- Q3(same as Q2), ssb-1g, dop=8
SELECT max(wx) FROM (
SELECT avg(LO_TAX) OVER (PARTITION BY C_REGION,C_NATION,C_CITY,S_CITY,S_NATION,S_REGION,P_TYPE ORDER BY LO_ORDERKEY ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING) AS wx FROM lineorder_flat
)a;
```

|  | query time | analytic time | partition search time |
|:--|:--|:--|:--|
| Q1-Before | 8.219s | 3s899ms | 733.219ms |
| Q1-After | 5.695s | 2s625ms | 48.535ms |
| Q2-Before | 17.267s | 6s115ms | 3s284ms |
| Q2-After | 13.056s | 2s94ms | 394.552ms |
| Q3-Before | 2.199s | 773.938ms | 385.516ms |
| Q3-After | 1.770s | 279.774ms | 49.531ms |

### CODE REVIEW

**Review by kangkaisen — DISMISSED** [review]

**Review by trueeyu — APPROVED** [review]

File: be/src/exec/vectorized/analytor.cpp [hunk 1]
```diff
@@ -547,9 +562,12 @@ bool Analytor::find_and_check_peer_group_end(bool is_found_partition_end_genuine
     _found_peer_group_end = _found_partition_end;
```
Comment (trueeyu) [inline be/src/exec/vectorized/analytor.cpp:562]: Could the `find_peer_group_end` use the algorithm?

Comment (liuyehcf) [reply be/src/exec/vectorized/analytor.cpp:562]: Yes, exactly, next pr~

File: be/src/exec/vectorized/analytor.cpp [hunk 2]
```diff
@@ -666,6 +687,23 @@ int64_t Analytor::_find_first_not_equal(vectorized::Column* column, int64_t targ
...
 
+void Analytor::_find_candidate_partition_ends() {
+    if (!_partition_statistics.is_high_cardinality()) {
+        return;
+    }
+
+    SCOPED_TIMER(_partition_search_timer);
+    for (size_t i = _found_partition_end + 1; i < _partition_columns[0]->size(); ++i) {
+        for (auto& column : _partition_columns) {
+            auto cmp = column->compare_at(i - 1, i, *column, 1);
```
Comment (mofeiatwork) [inline be/src/exec/vectorized/analytor.cpp:698]: Why not further optimize this compare procedure into a column-wise way to eliminate the `compare_at` virtual function call?

For example, compare the first column, to get a `compare_vector`, in which `0` denotes not equal and `1` denotes equal. Then compare the second column for equal elements only.

The `build_tie_for_column` and `ColumnTieBuilder` could save some work for you.

Comment (liuyehcf) [reply be/src/exec/vectorized/analytor.cpp:698]: I try to use tie, and find that it is much more slower than the simple loop.

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**